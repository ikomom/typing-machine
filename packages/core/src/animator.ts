import { calculatePatches, diffString, sliceInput } from './index'
import type { AnimatorStep, Patch } from './index'

export function* animatePatches(input: string, patches: Patch[]): Generator<AnimatorStep> {
  let content = input
  // 当前游标
  let cursor = 0
  // 每次只关注添加、删除的位置, 并且拼接起来
  for (let index = 0; index < patches.length; index++) {
    const patch = patches[index]
    const head = content.slice(0, patch.cursor)

    yield { type: 'new-patch', patch, index }

    if (patch.type === 'insert') {
      // 光标从头部插入
      cursor = patch.cursor
      const tail = content.slice(patch.cursor)
      // 模拟动画过程
      yield { type: 'insert-start', cursor, content }
      for (const { char, output, cursor: delta } of animateInsertionSlices(patch.content)) {
        yield {
          type: 'insert',
          cursor: cursor + delta,
          content: head + output + tail,
          char,
        }
      }
      content = head + patch.content + tail
      yield { type: 'insert-end', cursor, content }
    }
    else if (patch.type === 'removal') {
      // 光标从尾部开始减小
      cursor = patch.cursor + patch.length
      const tail = content.slice(patch.cursor + patch.length)
      const selection = content.slice(patch.cursor, patch.cursor + patch.length)
      yield { type: 'removal-start', cursor, content }

      for (let i = selection.length - 1; i >= 0; i--) {
        yield {
          type: 'removal',
          cursor: cursor - (selection.length - i),
          content: head + selection.slice(0, i) + tail,
        }
      }
      content = head + tail
      yield { type: 'removal-end', cursor, content }
    }
  }

  yield { type: 'animator-finish', content }
}

export function *animateInsertionSlices(input: string) {
  const slices = sliceInput(input)
  let output = ''
  for (const { content, cursor } of slices) {
    const head = output.slice(0, cursor)
    const tail = output.slice(cursor)

    let body = ''
    for (const char of content) {
      body += char
      // console.log('body', { body, char })
      yield {
        char,
        output: head + body + tail,
        cursor: cursor + body.length,
      }
    }
    output = head + content + tail
  }
}

export function animateTo(input: string, output: string) {
  const delta = diffString(input, output)
  const patches = calculatePatches(delta)
  return animatePatches(input, patches)
}

/**
 * TODO: 用法成疑
 * @param input
 * @param patches
 */
export function applyPatches(input: string, patches: Patch[]) {
  // 每次只关注添加、删除的位置, 并且拼接起来
  for (const patch of animatePatches(input, patches)) {
    if (patch.type === 'animator-finish')
      return patch.content
  }
  return input
}
