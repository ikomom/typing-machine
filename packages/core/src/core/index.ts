export * from './patch'
export * from './animator'
export * from './types'
export * from './snaps'
