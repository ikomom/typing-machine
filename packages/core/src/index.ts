export * from './patch'
export * from './animator'
export * from './types'
export * from './snaps'
export * from './utils'
export * from './slicing'
export * from './timming'
