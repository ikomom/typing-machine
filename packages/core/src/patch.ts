import { DiffMatchPatch, DiffOperation } from 'diff-match-patch-typescript'
import type { Diff } from 'diff-match-patch-typescript'
import type { Patch } from './types'

export function diffString(a: string, b: string): Diff[] {
  const diff = new DiffMatchPatch()
  const res = diff.diff_main(a, b)
  diff.diff_cleanupSemantic(res)
  return res
}

export function calculatePatches(diffs: Diff[]): Patch[] {
  const patches: Patch[] = []
  // 当前文字的index
  let cursor = 0
  for (const change of diffs) {
    const [operation, content] = change
    const offset = content.length
    if (operation === DiffOperation.DIFF_EQUAL) {
      cursor += offset
    }
    else if (operation === DiffOperation.DIFF_DELETE) {
      patches.push({
        type: 'removal',
        cursor,
        length: offset,
      })
      // 删除时，游标还是一样，所以不用移动，一下代码不用
      // cursor -= offset
    }
    else if (operation === DiffOperation.DIFF_INSERT) {
      patches.push({
        type: 'insert',
        cursor,
        content,
      })
      // 添加时, 游标 = 之前游标 + 添加长度
      cursor += offset
    }
    else {
      throw new Error('unknown change type')
    }
  }
  return patches
}
