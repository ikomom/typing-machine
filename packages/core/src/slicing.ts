import type { Slice } from './types'

const pairs = [
  ['{', '}'],
  ['(', ')'],
  ['[', ']'],
  ['<', '>'],
  ['`', '`'],
  ['\'', '\''],
  ['"', '"'],
]
const chars = '(){}\\[\\]<>\'"`'
const splitRegex = new RegExp(`(\s*[${chars}]\s*)`)

/**
 * slide input by pairs
 * @param input
 */
export function sliceInput(input: string): Slice[] {
  if (!input)
    return []
  // const rest = input.split(/\r?\n/).join('')
  const tailingNewLines = input.match(/\n+$/m)?.[0] || ''
  const rest = tailingNewLines.length
    ? input.slice(0, -tailingNewLines.length)
    : input

  // string slice
  const strings = rest.split(splitRegex).filter(Boolean)

  let index = 0
  const slices = strings
    .map((s, idx) => {
      const trimmed = s.trim()
      const pair = pairs.find(i => i[1] === trimmed)
      // pair before 2 index
      const order = (pair && strings[idx - 2]?.trim() === pair[0]) ? idx - 2 : idx

      const _index = index
      index += s.length

      return {
        content: s,
        order,
        cursor: _index,
      }
    }).sort((a, b) => a.order - b.order)

  let cursor = 0
  slices.forEach((i) => {
    i.cursor = Math.min(i.cursor, cursor)
    cursor += i.content.length
  })
  if (tailingNewLines.length) {
    slices.unshift({
      content: tailingNewLines,
      order: -1,
      cursor: 0,
    })
  }
  // console.log({ slices: [...slices], strings })
  return slices
}
