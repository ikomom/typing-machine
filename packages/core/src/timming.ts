export const timeRanges: [string, number, number][] = [
  // default
  ['', 0, 100],
  // commons
  [' ', 0, 40],
  ['adefijklosu', 0, 50],
  ['bcghmnpqrtvwxyz', 5, 70],
  // uppercase
  ['adefijklosu'.toUpperCase(), 0, 50],
  ['bcghmnpqrtvwxyz'.toUpperCase(), 5, 70],
  // a bit far
  ['[];\',./-=', 10, 120],
  // need shift
  ['<>?:"{}|\\_+()*', 20, 140],
  // a bit far and shift
  ['!~`#$%^&', 20, 180],
]

export function randRange(min: number, max: number): number {
  return Math.random() * (max - min) + min
}

export function getTimeout(char: string, multiplier = 1): number {
  const [, min, max] = timeRanges.find(range => range[0].includes(char)) || timeRanges[0]
  return randRange(min, max) * multiplier
}
