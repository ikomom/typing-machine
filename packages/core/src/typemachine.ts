import { getTimeout, randRange } from './timing'
import type { AnimatorStep } from './types'

export function sleep(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms))
}

export interface TypewriterOptions {

}

export async function *typingAnimator(
  steps: Generator<AnimatorStep>,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  options: TypewriterOptions = {},
): AsyncGenerator<AnimatorStep> {
  for (const step of steps) {
    switch (step.type) {
      case 'init':
        break
      case 'new-snap':
        if (step.index)
          await sleep(randRange(700, 1000))
        break
      case 'new-patch':
        if (step.index)
          await sleep(randRange(200, 500))
        break
      case 'insert':
        await sleep(getTimeout(step.char, 1.2))
        break
      case 'removal':
        await sleep(randRange(0, 5))
        break
    }
    yield step
  }
}
