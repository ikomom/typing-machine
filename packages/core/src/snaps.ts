import { simpleAnimator } from './animator'
import type { AnimatorStep, Snapshot } from './types'
import { replaceAll } from './utils'

export const SNAP_HEADING = 'ik-typing-machine Snapshots v1'
export const SNAP_SEPARATOR_PRE = '-'.repeat(2)
export const SNAP_SEPARATOR_SUFFIX = '-'.repeat(10)
export const SNAP_SEPARATOR_OPTIONS = '----OPTIONS----'
export const SNAP_SEPARATOR = `${SNAP_SEPARATOR_PRE}--${SNAP_SEPARATOR_SUFFIX}`
export const snapSeparatorMatcher = (eol: string) => new RegExp(`${eol}?${SNAP_SEPARATOR_PRE}[#\\w-]*${SNAP_SEPARATOR_SUFFIX}${eol}`, 'g')
export const snapSeparatorMatcherOptions = (eol: string) => new RegExp(`${eol}?${SNAP_SEPARATOR_OPTIONS}`, 'g')

/**
 * Snapshot store
 */
export class Snapshots extends Array<Snapshot> {
  constructor(...props: Snapshot[]) {
    super(...props)
  }

  get last() {
    return this[this.length - 1]
  }

  get first() {
    return this[0]
  }

  toString(eol = '\n'): string {
    return [
      SNAP_HEADING + eol,
      ...this.flatMap((snap, i) => {
        return [
          `${SNAP_SEPARATOR_PRE + String(i + 1).padStart(2, '0') + SNAP_SEPARATOR_SUFFIX}`,
          snap.content,
          ...(snap.options
            ? [
                SNAP_SEPARATOR_OPTIONS,
                Object.keys(snap.options).length > 1
                  ? JSON.stringify(snap.options, null, 2)
                  : JSON.stringify(snap.options)]
            : []),
        ]
      }),
      SNAP_SEPARATOR,
      '',
    ].join(eol)
  }

  static fromRawStr(raw: string) {
    const parts = raw.split('\r?\n')
  }

  static fromString(raw: string, eol = '\n') {
    let parts = raw
      .split(snapSeparatorMatcher(eol))
      // .slice(1, -1) // remove header and tailing
    // console.log({ raw, q: replaceAll(raw, eol, '\n'), eol, parts })
    parts = parts.slice(1, -1)
    const snapshots: Snapshot[] = []
    for (let i = 0; i < parts.length; i += 1) {
      const [content, optionsRaw] = parts[i].split(snapSeparatorMatcherOptions(eol))
      const snap: Snapshot = {
        content,
      }
      if (optionsRaw?.trim())
        snap.options = JSON.parse(optionsRaw)

      snapshots.push(snap)
    }

    return new Snapshots(...snapshots)
  }

  *animate(): Generator<AnimatorStep> {
    let lastContent: string | undefined
    const copy: Snapshot[] = [...this]

    for (let index = 0; index < copy.length; index++) {
      const snap = copy[index]
      if (lastContent === undefined) {
        lastContent = snap.content
        yield {
          type: 'init',
          content: lastContent,
        }
        continue
      }

      yield {
        type: 'new-snap',
        snap,
        index,
      }
      // form last to current, generate animator array
      const animator = simpleAnimator(lastContent, snap.content)
      for (const result of animator)
        yield result

      lastContent = snap.content
    }
  }
}

export type SnapshotFallbackLoader = (id: string) => Snapshots | undefined | Promise<Snapshots | undefined>
export interface SnapshotManagerOptions {
  ensureFallback?: SnapshotFallbackLoader
}

export class SnapshotManager extends Map<string, Snapshots> {
  constructor(
    public options: SnapshotManagerOptions,
  ) {
    super()
  }

  /**
   * set snapshot by id, if no present, read load callback or create new snapshot
   * @param id
   * @param load
   */
  async ensure(id: string, load = this.options.ensureFallback): Promise<Snapshots> {
    if (!this.has(id))
      this.set(id, await load?.(id) || new Snapshots())

    return this.get(id)!
  }
}
