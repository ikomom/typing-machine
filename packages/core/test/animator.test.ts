import { describe, expect, it } from 'vitest'
import { simpleAnimator } from '../src'
import { input, output } from './fixture'

describe('animator', () => {
  it('generate', () => {
    const animator = simpleAnimator(input, output)
    expect([...animator]).toMatchSnapshot('animator')
  })
})
