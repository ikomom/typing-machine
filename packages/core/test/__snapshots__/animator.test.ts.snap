// Vitest Snapshot v1

exports[`animator > generate > animator 1`] = `
[
  {
    "cursor": 71,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} from './one'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 0,
    "type": "removal-start",
  },
  {
    "cursor": 70,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} from './one

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 24,
    "type": "removal-ing",
  },
  {
    "cursor": 69,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} from './on

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 23,
    "type": "removal-ing",
  },
  {
    "cursor": 68,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} from './o

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 22,
    "type": "removal-ing",
  },
  {
    "cursor": 67,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} from './

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 21,
    "type": "removal-ing",
  },
  {
    "cursor": 66,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} from '.

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 20,
    "type": "removal-ing",
  },
  {
    "cursor": 65,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} from '

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 19,
    "type": "removal-ing",
  },
  {
    "cursor": 64,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} from 

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 18,
    "type": "removal-ing",
  },
  {
    "cursor": 63,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} from

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 17,
    "type": "removal-ing",
  },
  {
    "cursor": 62,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} fro

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 16,
    "type": "removal-ing",
  },
  {
    "cursor": 61,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} fr

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 15,
    "type": "removal-ing",
  },
  {
    "cursor": 60,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} f

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 14,
    "type": "removal-ing",
  },
  {
    "cursor": 59,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} 

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 13,
    "type": "removal-ing",
  },
  {
    "cursor": 58,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll}

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 12,
    "type": "removal-ing",
  },
  {
    "cursor": 57,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 11,
    "type": "removal-ing",
  },
  {
    "cursor": 56,
    "output": "
import { describe, expect, it } from 'vitest'
import {l

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 10,
    "type": "removal-ing",
  },
  {
    "cursor": 55,
    "output": "
import { describe, expect, it } from 'vitest'
import {

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 9,
    "type": "removal-ing",
  },
  {
    "cursor": 54,
    "output": "
import { describe, expect, it } from 'vitest'
import 

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 8,
    "type": "removal-ing",
  },
  {
    "cursor": 53,
    "output": "
import { describe, expect, it } from 'vitest'
import

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 7,
    "type": "removal-ing",
  },
  {
    "cursor": 52,
    "output": "
import { describe, expect, it } from 'vitest'
impor

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 6,
    "type": "removal-ing",
  },
  {
    "cursor": 51,
    "output": "
import { describe, expect, it } from 'vitest'
impo

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 5,
    "type": "removal-ing",
  },
  {
    "cursor": 50,
    "output": "
import { describe, expect, it } from 'vitest'
imp

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 4,
    "type": "removal-ing",
  },
  {
    "cursor": 49,
    "output": "
import { describe, expect, it } from 'vitest'
im

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 3,
    "type": "removal-ing",
  },
  {
    "cursor": 48,
    "output": "
import { describe, expect, it } from 'vitest'
i

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 2,
    "type": "removal-ing",
  },
  {
    "cursor": 47,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 1,
    "type": "removal-ing",
  },
  {
    "cursor": 46,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 0,
    "type": "removal-ing",
  },
  {
    "cursor": 71,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 46,
      "length": 25,
      "type": "removal",
    },
    "patchIndex": 0,
    "type": "removal-end",
  },
  {
    "cursor": 138,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 1,
    "type": "removal-start",
  },
  {
    "cursor": 137,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello)
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 23,
    "type": "removal-ing",
  },
  {
    "cursor": 136,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hell)
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 22,
    "type": "removal-ing",
  },
  {
    "cursor": 135,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hel)
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 21,
    "type": "removal-ing",
  },
  {
    "cursor": 134,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('he)
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 20,
    "type": "removal-ing",
  },
  {
    "cursor": 133,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('h)
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 19,
    "type": "removal-ing",
  },
  {
    "cursor": 132,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log(')
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 18,
    "type": "removal-ing",
  },
  {
    "cursor": 131,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log()
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 17,
    "type": "removal-ing",
  },
  {
    "cursor": 130,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log)
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 16,
    "type": "removal-ing",
  },
  {
    "cursor": 129,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.lo)
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 15,
    "type": "removal-ing",
  },
  {
    "cursor": 128,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.l)
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 14,
    "type": "removal-ing",
  },
  {
    "cursor": 127,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.)
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 13,
    "type": "removal-ing",
  },
  {
    "cursor": 126,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console)
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 12,
    "type": "removal-ing",
  },
  {
    "cursor": 125,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    consol)
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 11,
    "type": "removal-ing",
  },
  {
    "cursor": 124,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    conso)
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 10,
    "type": "removal-ing",
  },
  {
    "cursor": 123,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    cons)
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 9,
    "type": "removal-ing",
  },
  {
    "cursor": 122,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    con)
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 8,
    "type": "removal-ing",
  },
  {
    "cursor": 121,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    co)
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 7,
    "type": "removal-ing",
  },
  {
    "cursor": 120,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    c)
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 6,
    "type": "removal-ing",
  },
  {
    "cursor": 119,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    )
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 5,
    "type": "removal-ing",
  },
  {
    "cursor": 118,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
   )
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 4,
    "type": "removal-ing",
  },
  {
    "cursor": 117,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
  )
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 3,
    "type": "removal-ing",
  },
  {
    "cursor": 116,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
 )
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 2,
    "type": "removal-ing",
  },
  {
    "cursor": 115,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
)
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 1,
    "type": "removal-ing",
  },
  {
    "cursor": 114,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33))
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 0,
    "type": "removal-ing",
  },
  {
    "cursor": 138,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33))
  })
})
",
    "patch": {
      "from": 114,
      "length": 24,
      "type": "removal",
    },
    "patchIndex": 1,
    "type": "removal-end",
  },
  {
    "cursor": 114,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33))
  })
})
",
    "patch": {
      "from": 114,
      "text": ".toEqual(1",
      "type": "insert",
    },
    "patchIndex": 2,
    "type": "insert-start",
  },
  {
    "char": ".",
    "cursor": 115,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33).)
  })
})
",
    "patch": {
      "from": 114,
      "text": ".toEqual(1",
      "type": "insert",
    },
    "patchIndex": 2,
    "type": "insert-ing",
  },
  {
    "char": "t",
    "cursor": 116,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33).t)
  })
})
",
    "patch": {
      "from": 114,
      "text": ".toEqual(1",
      "type": "insert",
    },
    "patchIndex": 2,
    "type": "insert-ing",
  },
  {
    "char": "o",
    "cursor": 117,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33).to)
  })
})
",
    "patch": {
      "from": 114,
      "text": ".toEqual(1",
      "type": "insert",
    },
    "patchIndex": 2,
    "type": "insert-ing",
  },
  {
    "char": "E",
    "cursor": 118,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33).toE)
  })
})
",
    "patch": {
      "from": 114,
      "text": ".toEqual(1",
      "type": "insert",
    },
    "patchIndex": 2,
    "type": "insert-ing",
  },
  {
    "char": "q",
    "cursor": 119,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33).toEq)
  })
})
",
    "patch": {
      "from": 114,
      "text": ".toEqual(1",
      "type": "insert",
    },
    "patchIndex": 2,
    "type": "insert-ing",
  },
  {
    "char": "u",
    "cursor": 120,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33).toEqu)
  })
})
",
    "patch": {
      "from": 114,
      "text": ".toEqual(1",
      "type": "insert",
    },
    "patchIndex": 2,
    "type": "insert-ing",
  },
  {
    "char": "a",
    "cursor": 121,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33).toEqua)
  })
})
",
    "patch": {
      "from": 114,
      "text": ".toEqual(1",
      "type": "insert",
    },
    "patchIndex": 2,
    "type": "insert-ing",
  },
  {
    "char": "l",
    "cursor": 122,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33).toEqual)
  })
})
",
    "patch": {
      "from": 114,
      "text": ".toEqual(1",
      "type": "insert",
    },
    "patchIndex": 2,
    "type": "insert-ing",
  },
  {
    "char": "(",
    "cursor": 123,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33).toEqual()
  })
})
",
    "patch": {
      "from": 114,
      "text": ".toEqual(1",
      "type": "insert",
    },
    "patchIndex": 2,
    "type": "insert-ing",
  },
  {
    "char": "1",
    "cursor": 124,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33).toEqual(1)
  })
})
",
    "patch": {
      "from": 114,
      "text": ".toEqual(1",
      "type": "insert",
    },
    "patchIndex": 2,
    "type": "insert-ing",
  },
  {
    "cursor": 114,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33).toEqual(1)
  })
})
",
    "patch": {
      "from": 114,
      "text": ".toEqual(1",
      "type": "insert",
    },
    "patchIndex": 2,
    "type": "insert-end",
  },
]
`;
